src/platform/android/DnssdImpl.cpp

```diff
@@ -242,8 +264,79 @@ void HandleResolve(jstring instanceName, jstring serviceType, jstring address, j
     CopyString(service.mName, jniInstanceName.c_str());
     CopyString(service.mType, jniServiceType.c_str());
     service.mPort = static_cast<uint16_t>(port);
+    service.mTextEntrySize = 0;
+    service.mTextEntries   = nullptr;
+
+    // Note on alloc/free memory use
+    // We are only allocating the entries list and the data field of each entry
+    // so we free these in the exit section
+    if (textEntries != nullptr)
+    {
+        jobjectArray keys   = (jobjectArray) env->CallObjectMethod(sMdnsCallbackObject, sGetTextEntryKeysMethod, textEntries);
+        size_t size         = env->GetArrayLength(keys);
+        TextEntry * entries = new (std::nothrow) TextEntry[size];
+        if (entries == nullptr)
+        {
+            ChipLogError(Discovery, "entries alloc failure");
+            goto exit;
+        }
+        service.mTextEntries = entries;
+        for (size_t i = 0; i < size; i++)
+        {
+            jstring jniKeyObject = (jstring) env->GetObjectArrayElement(keys, i);
+            JniUtfString key(env, jniKeyObject);
+            entries[i].mKey = strdup(key.c_str());
+
+            jbyteArray datas =
+                (jbyteArray) env->CallObjectMethod(sMdnsCallbackObject, sGetTextEntryDataMethod, textEntries, jniKeyObject);
+            if (datas != nullptr)
+            {
+                size_t dataSize = env->GetArrayLength(datas);
+                uint8_t * data  = new (std::nothrow) uint8_t[dataSize];
+                if (data == nullptr)
+                {
+                    ChipLogError(Discovery, "data alloc failure");
+                    goto exit;
+                }
+
+                jbyte * jnidata = env->GetByteArrayElements(datas, nullptr);
+                for (size_t j = 0; j < dataSize; j++)
+                {
+                    data[j] = static_cast<uint8_t>(jnidata[j]);
+                }
+                entries[i].mDataSize = dataSize;
+                entries[i].mData     = data;
+
+                ChipLogProgress(Discovery, " ----- entry [%u] : %s %s\n", static_cast<unsigned int>(i), entries[i].mKey,
+                                std::string(reinterpret_cast<char *>(data), dataSize).c_str());
+            }
+            else
+            {
+                ChipLogProgress(Discovery, " ----- entry [%u] : %s NULL\n", static_cast<unsigned int>(i), entries[i].mKey);
+
+                entries[i].mDataSize = 0;
+                entries[i].mData     = nullptr;
+            }
+            service.mTextEntrySize = size;
+        }
+    }
 
+exit:
     dispatch(CHIP_NO_ERROR, &service, &ipAddress);
+
+    if (service.mTextEntries != nullptr)
+    {
+        size_t size = service.mTextEntrySize;
+        for (size_t i = 0; i < size; i++)
+        {
+            delete[] service.mTextEntries[i].mKey;
```

Comment:
I think strdup is supposed to be paired with free()

Reply:
can't use free on const char *